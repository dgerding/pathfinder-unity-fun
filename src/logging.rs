use std::path::PathBuf;
use std::fs::{File, OpenOptions};
use std::io::prelude::*;
use std::sync::{Mutex, Once};
use std::env;
use log::{Record, LevelFilter, Metadata};

static INIT: Once = Once::new();

static mut GLOBAL_LOGGER: Option<Logger> = None;

struct Logger {
    logfile: PathBuf,
    mutex: Mutex<()>
}

impl Logger {
    pub fn new(logfile: PathBuf) -> Self {
        Logger { logfile, mutex: Mutex::new(()) }
    }

    pub fn log<T: AsRef<str>>(&mut self, msg: T) {
        // Since Unity's renderer may be multi-threaded, who
        // knows where we might be called from, so we'll use a mutex just in case.
        let _ = self.mutex.lock().unwrap();

        if !self.logfile.exists() {
            File::create(&self.logfile).unwrap();
        }
        let mut file = OpenOptions::new().append(true).open(&self.logfile).unwrap();
        file.write(msg.as_ref().as_bytes()).unwrap();
        file.write(b"\n").unwrap();
        file.flush().unwrap();
    }
}

fn log_raw<T: AsRef<str>>(msg: T) {
    INIT.call_once(|| {
        let mut logfile = env::current_dir().unwrap();
        logfile.push("pathfinder-plugin.log");
        let logger = Logger::new(logfile);
        unsafe {
            GLOBAL_LOGGER = Some(logger);
        }
    });
    if let Some(logger) = unsafe { &mut GLOBAL_LOGGER } {
        logger.log(msg);
    } else {
        panic!("Expected logger to exist!");
    }
}

// This code integrates with the "log" crate: https://docs.rs/log/0.4.6/log/
// It was written after the original logging code, which is why the
// architecture here is kind of funky.

struct LogWrapper;

static LOG_LOGGER: LogWrapper = LogWrapper;

impl log::Log for LogWrapper {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        log_raw(format!("{}", record.args()));
    }

    fn flush(&self) {}
}

pub fn init() {
    log::set_logger(&LOG_LOGGER)
      .map(|()| log::set_max_level(LevelFilter::Info)).unwrap();
}
