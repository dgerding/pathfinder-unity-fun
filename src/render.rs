use std::path::PathBuf;
use pathfinder_canvas::{CanvasRenderingContext2D};
use pathfinder_geometry::basic::vector::Vector2I;
use pathfinder_gpu::resources::FilesystemResourceLoader;
use pathfinder_gl::{GLDevice, GLVersion};
use pathfinder_renderer::gpu::renderer::DestFramebuffer;
use pathfinder_renderer::concurrent::rayon::RayonExecutor;
use pathfinder_renderer::concurrent::scene_proxy::SceneProxy;
use pathfinder_renderer::options::RenderOptions;
use pathfinder_renderer::gpu::renderer::Renderer as PathfinderRenderer;
use gl::types::GLuint;

use crate::gl_util::{get_viewport_size, get_draw_framebuffer_binding};

fn get_current_window_size() -> Vector2I {
    let (_, _, width, height) = get_viewport_size();
    Vector2I::new(width, height)
}

fn build_renderer(
    window_size: Vector2I,
    framebuffer: GLuint,
    loader: &FilesystemResourceLoader
) -> PathfinderRenderer<GLDevice> {
    pathfinder_renderer::gpu::renderer::Renderer::new(
        GLDevice::new(GLVersion::GL3, framebuffer),
        loader,
        DestFramebuffer::full_window(window_size)
    )
}

pub struct Renderer {
    renderer: PathfinderRenderer<GLDevice>,
    window_size: Vector2I,
    framebuffer: GLuint
}

impl Renderer {
    pub fn new(resources_dir: PathBuf) -> Self {
        let window_size = get_current_window_size();
        let framebuffer = get_draw_framebuffer_binding();
        let loader = FilesystemResourceLoader { directory: resources_dir };
        let renderer = build_renderer(window_size, framebuffer, &loader);

        Renderer { renderer, window_size, framebuffer }
    }

    // If Unity's window size/framebuffer changes, make sure our draw
    // calls adapt.
    fn sync_gfx_state(&mut self) {
        let framebuffer = get_draw_framebuffer_binding();
        let window_size = get_current_window_size();
        if window_size != self.window_size || framebuffer != self.framebuffer {
            info!(
                "Window size/framebuffer changed from {:?}/{} to {:?}/{}.",
                self.window_size,
                self.framebuffer,
                window_size,
                framebuffer
            );
            self.window_size = window_size;
            self.framebuffer = framebuffer;
            self.renderer.device.set_default_framebuffer(framebuffer);
            self.renderer.replace_dest_framebuffer(DestFramebuffer::full_window(window_size));
            self.renderer.set_main_framebuffer_size(window_size);
        }
    }

    pub fn render(&mut self, canvas: Box<CanvasRenderingContext2D>) {
        self.sync_gfx_state();
        let renderer = &mut self.renderer;

        // Render the canvas to screen.
        let scene = SceneProxy::from_scene(canvas.into_scene(), RayonExecutor);
        scene.build_and_render(renderer, RenderOptions::default());
    }
}
