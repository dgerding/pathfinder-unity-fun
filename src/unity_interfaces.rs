// This is essentially a Rust version of the following header files:
//
// https://bitbucket.org/Unity-Technologies/graphicsdemos/src/default/NativeRenderingPlugin/PluginSource/source/Unity/IUnityInterface.h
// https://bitbucket.org/Unity-Technologies/graphicsdemos/src/default/NativeRenderingPlugin/PluginSource/source/Unity/IUnityGraphics.h

use libc::{
    c_ulonglong,
    c_int,
    c_void
};

#[repr(C)]
pub struct UnityInterfaceGUID {
    pub high: c_ulonglong,
    pub low: c_ulonglong,
}

// Type alias to indicate that we don't need a struct member, so we won't bother writing
// a (potentially incorrect) typing for it. However, it does occupy the size of a C void
// pointer in a struct, so we want to declare it for memory layout purposes.
type Unneeded = *const c_void;

#[repr(C)]
pub struct IUnityInterfaces {
    // I have no idea how passing a struct by value works in FFI, so I'm not going to use
    // the non-split versions of the API.
    get_interface: Unneeded,
    register_interface: Unneeded,

    get_interface_split: extern "stdcall" fn(guid_high: c_ulonglong, guid_low: c_ulonglong) -> *const c_void,
    register_interface_split: Unneeded,
}

impl IUnityInterfaces {
    pub fn get_unity_graphics(&self) -> *const IUnityGraphics {
        (self.get_interface_split)(I_UNITY_GRAPHICS_GUID.high, I_UNITY_GRAPHICS_GUID.low) as *const IUnityGraphics
    }
}

#[derive(Debug)]
pub enum UnityGfxRenderer {
    // OpenGL            =  0, // Legacy OpenGL, removed
    // D3D9              =  1, // Direct3D 9, removed
    D3D11             =  2, // Direct3D 11
    GCM               =  3, // PlayStation 3
    Null              =  4, // "null" device (used in batch mode)
    OpenGLES20        =  8, // OpenGL ES 2.0
    OpenGLES30        = 11, // OpenGL ES 3.0
    GXM               = 12, // PlayStation Vita
    PS4               = 13, // PlayStation 4
    XboxOne           = 14, // Xbox One
    Metal             = 16, // iOS Metal
    OpenGLCore        = 17, // OpenGL core
    D3D12             = 18, // Direct3D 12
    Vulkan            = 21, // Vulkan
    Nvn               = 22, // Nintendo Switch NVN API
    XboxOneD3D12      = 23  // MS XboxOne Direct3D 12
}

#[derive(Debug)]
pub enum UnityGfxDeviceEventType
{
    Initialize     = 0,
    Shutdown       = 1,
    BeforeReset    = 2,
    AfterReset     = 3,
}

type IUnityGraphicsDeviceEventCallback = extern "stdcall" fn(event_type: UnityGfxDeviceEventType);

pub struct IUnityGraphics {
    pub get_renderer: extern "stdcall" fn() -> UnityGfxRenderer,
    pub register_device_event_callback: extern "stdcall" fn(cb: IUnityGraphicsDeviceEventCallback),
    pub unregister_device_event_callback: extern "stdcall" fn(cb: IUnityGraphicsDeviceEventCallback),
    pub reserve_event_id_range: extern "stdcall" fn(count: c_int) -> c_int
}

pub const I_UNITY_GRAPHICS_GUID: UnityInterfaceGUID = UnityInterfaceGUID {
    high: 0x7CBA0A9CA4DDB544,
    low: 0x8C5AD4926EB17B11
};
