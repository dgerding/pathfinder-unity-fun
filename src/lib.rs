#[macro_use]
extern crate enum_primitive_derive;

use std::env;
use std::path::PathBuf;
use libc::c_int;

mod unity_interfaces;
mod gl_util;
mod render;
mod logging;

use unity_interfaces::{
    IUnityGraphics,
    IUnityInterfaces,
    UnityGfxRenderer,
    UnityRenderingEvent,
    UnityGfxDeviceEventType,
    UnityGfxDeviceEventTypeInt
};
use render::Renderer;
use logging::log;

struct PluginState {
    unity_interfaces: *const IUnityInterfaces,
    unity_renderer: Option<UnityGfxRenderer>,
    renderer: Option<Renderer>,
    errored: bool
}

impl PluginState {
    pub fn new(unity_interfaces: *const IUnityInterfaces) -> Self {
        let mut plugin = PluginState {
            unity_interfaces,
            unity_renderer: None,
            renderer: None,
            errored: false
        };
        plugin.initialize();
        plugin
    }

    fn find_resources_dir(&mut self) -> Option<PathBuf> {
        let mut resources_dir = env::current_dir().unwrap();
        resources_dir.push("unity-project_Data");
        resources_dir.push("StreamingAssets");
        resources_dir.push("pathfinder");
        log(format!("Searching for resources at {}.", resources_dir.to_string_lossy()));
        if resources_dir.exists() {
            Some(resources_dir)
        } else {
            // TODO: Also look in the "Assets" folder, if we're being run
            // inside the Unity editor?
            None
        }
    }

    fn initialize(&mut self) {
        log("Pathfinder plugin initialized.");
        unsafe {
            self.log_unity_renderer_info();
            let gfx = self.get_unity_graphics();
            ((*gfx).register_device_event_callback)(handle_unity_device_event);
        }
    }

    fn get_unity_graphics(&self) -> *const IUnityGraphics {
        unsafe {
            let gfx = (*self.unity_interfaces).get_unity_graphics();
            assert!(!gfx.is_null());
            gfx
        }
    }

    pub fn get_unity_renderer(&self) -> Option<UnityGfxRenderer> {
        let gfx = self.get_unity_graphics();
        unsafe {
            ((*gfx).get_renderer)().convert()
        }
    }

    pub fn log_unity_renderer_info(&mut self) {
        log(format!("Unity renderer is {:?}.", self.get_unity_renderer()));
    }

    pub fn handle_unity_device_event(&mut self, event_type: Option<UnityGfxDeviceEventType>) {
        log(format!("Unity graphics event occurred: {:?}", event_type));
        match event_type {
            Some(UnityGfxDeviceEventType::Initialize) => {
                self.log_unity_renderer_info();
                self.unity_renderer = self.get_unity_renderer();
                if let Some(UnityGfxRenderer::OpenGLCore) = self.unity_renderer {
                    gl_util::init();
                    let (major, minor) = gl_util::get_version();
                    let version = gl_util::get_version_string();
                    log(format!("OpenGL version is {}.{} ({}).", major, minor, version));
                }
            },
            _ => {}
        }
    }

    fn try_to_init_renderer(&mut self) {
        match self.find_resources_dir() {
            None => {
                log("Unable to find resources dir.");
                self.errored = true;
            },
            Some(resources_dir) => {
                self.renderer = Some(Renderer::new(resources_dir));
            }
        }
    }

    pub fn render(&mut self) {
        if self.errored {
            return;
        }
        if let Some(UnityGfxRenderer::OpenGLCore) = self.unity_renderer {
            if self.renderer.is_none() {
                self.try_to_init_renderer();
            }
            if let Some(renderer) = &mut self.renderer {
                renderer.render();
            }
        } else {
            log(format!(
                "render() called, but rendering backend {:?} is unsupported.",
                self.unity_renderer
            ));
            self.errored = true;
        }
    }
}

impl Drop for PluginState {
    fn drop(&mut self) {
        log("Shutting down Pathfinder plugin.");
        // TODO: We should ideally unregister our device event callback here, but
        // we never seem to actually get called, so I guess it's not that urgent...
    }
}

static mut PLUGIN_STATE: Option<PluginState> = None;

extern "stdcall" fn handle_unity_device_event(event_type: UnityGfxDeviceEventTypeInt) {
    get_plugin_state_mut().handle_unity_device_event(event_type.convert());
}

// This is called by Unity when the plugin is loaded.
#[no_mangle]
pub extern "stdcall" fn UnityPluginLoad(unity_interfaces: *const IUnityInterfaces) {
    unsafe {
        assert!(PLUGIN_STATE.is_none());
        let plugin = PluginState::new(unity_interfaces);
        PLUGIN_STATE = Some(plugin);
    }
}

// This is called by Unity when the plugin is about to be unloaded.
#[no_mangle]
pub extern "stdcall" fn UnityPluginUnload() {
    unsafe {
        assert!(PLUGIN_STATE.is_some());
        PLUGIN_STATE = None;
    }
}

fn get_plugin_state_mut() -> &'static mut PluginState {
    unsafe {
        match PLUGIN_STATE {
            None => panic!("Expected plugin to be initialized!"),
            Some(ref mut state) => state
        }
    }
}

// TODO: Remove this function, we don't need it. But also
// remove any C# code that calls into it.
#[no_mangle]
pub extern "stdcall" fn boop_stdcall(x: i32) -> i32 {
    log(format!("boop_stdcall({}) called.", x));
    51 + x
}

extern "stdcall" fn handle_render_event(_event_id: c_int) {
    get_plugin_state_mut().render();
}

#[no_mangle]
pub extern "stdcall" fn get_render_event_func() -> UnityRenderingEvent {
    handle_render_event
}
