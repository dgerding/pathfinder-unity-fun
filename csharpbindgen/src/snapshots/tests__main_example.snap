---
created: "2019-06-30T21:50:06.510808900Z"
creator: insta@0.8.1
source: src/lib.rs
expression: code
---
// This file has been auto-generated, please do not edit it.

using System;
using System.Runtime.InteropServices;

[Serializable]
[StructLayout(LayoutKind.Sequential)]
internal struct MyStruct {
    internal float foo;
    internal float bar;

    internal MyStruct(float foo, float bar) {
        this.foo = foo;
        this.bar = bar;
    }
}

internal class MyStuff {
    internal const byte BOOP = 1;

    [DllImport("MyDll")]
    internal static extern byte blarg(Int32 a, ref MyStruct b, IntPtr /* MyOpaqueStruct */ c);

}

